### PR TITLE
Add tests for generated plots of ``plot_edf``

### PR DESCRIPTION
**Pull request by makinzm**
## Motivation

Add test in matplotlib test ``plot_edf``  of plotly.

## Description of the changes

This test confirms 
- that the values are monotonically non-decreasing and  
- that the values are in [0,1].

## Reference
- [Stabilise visualisation module #2959](https://github.com/optuna/optuna/issues/2959)
- [Add tests for generated plots of matplotlib.plot_edf #3178](https://github.com/optuna/optuna/pull/3178)

### CODE REVIEW

File: tests/visualization_tests/test_edf.py [hunk 1]
```diff
@@ -28,27 +39,39 @@ def test_plot_optimization_history(direction: str) -> None:
     study0 = create_study(direction=direction)
     study0.optimize(lambda t: t.suggest_float("x", 0, 5), n_trials=10)
     figure = plot_edf(study0)
+    data = figure.data
+    _validate(data[0]["y"])
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_edf.py:43]: The `data` variable is only used in a single line. How about simplifying it as follows?
```suggestion
    _validate(figure.data[0]["y"])
```

File: tests/visualization_tests/test_edf.py [hunk 2]
```diff
@@ -28,27 +39,39 @@ def test_plot_optimization_history(direction: str) -> None:
...
     assert len(figure.data) == 1
     assert figure.layout.xaxis.title.text == "Objective Value"
 
     # Test with two studies.
     study1 = create_study(direction=direction)
     study1.optimize(lambda t: t.suggest_float("x", 0, 5), n_trials=10)
     figure = plot_edf([study0, study1])
+    data = figure.data
+    for tmp_data in data:
+        _validate(tmp_data["y"])
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_edf.py:53]: ditto. In addition, `tmp_data` may be a bit ambiguous. Can we rename it?
```suggestion
    for points in figure.data:
        _validate(points["y"])
```

File: tests/visualization_tests/test_edf.py [hunk 3]
```diff
@@ -28,27 +39,39 @@ def test_plot_optimization_history(direction: str) -> None:
...
     assert len(figure.data) == 2
     figure = plot_edf((study0, study1))
+    data = figure.data
+    for tmp_data in data:
+        _validate(tmp_data["y"])
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_edf.py:58]: ditto.
```suggestion
    for points in figure.data:
        _validate(points["y"])
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. The change basically looks good to me. I added some small comments.

File: tests/visualization_tests/test_edf.py [hunk 4]
```diff
@@ -11,6 +14,14 @@ def test_target_is_none_and_study_is_multi_obj() -> None:
         plot_edf(study)
 
 
+def _validate(y_values: Sequence[float]) -> None:
```
Comment (nzw0301) [inline tests/visualization_tests/test_edf.py:17]: `_validate` sounds unclearn. how about `_validate_y_values`?
```suggestion
def _validate_y_values(y_values: Sequence[float]) -> None:
```

File: tests/visualization_tests/test_edf.py [hunk 5]
```diff
@@ -11,6 +14,14 @@ def test_target_is_none_and_study_is_multi_obj() -> None:
...
+    np_values = np.array(y_values)
+    # This line confirms that the values are monotonically non-decreasing.
```
Comment (nzw0301) [inline tests/visualization_tests/test_edf.py:19]: how about this simplification?
```suggestion
    # Confirms that the values are monotonically non-decreasing.
```

File: tests/visualization_tests/test_edf.py [hunk 6]
```diff
@@ -11,6 +14,14 @@ def test_target_is_none_and_study_is_multi_obj() -> None:
...
+    assert np.all(np_values[1:] - np_values[:-1] >= 0)
+    # This line confirms that the values are in [0,1].
```
Comment (nzw0301) [inline tests/visualization_tests/test_edf.py:21]:
```suggestion

    # Confirms that the values are in [0,1].
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]

**Comment by makinzm** [comment]
Thank you for your modification.

File: tests/visualization_tests/test_edf.py [hunk 7]
```diff
@@ -11,6 +14,15 @@ def test_target_is_none_and_study_is_multi_obj() -> None:
         plot_edf(study)
 
 
+def _validate_edf_values(edf_values: Sequence[float]) -> None:
+    np_values = np.array(y_values)
+    # This line confirms that the values are monotonically non-decreasing.
+    assert np.all(np_values[1:] - np_values[:-1] >= 0)
+
+    # Confirms that the values are in [0,1].
+    assert np.all((0 <= np_values) * (np_values <= 1))
```
Comment (himkt) [inline tests/visualization_tests/test_edf.py:23]: `&` is recommended for logical and operation.
🔗 https://numpy.org/doc/stable/reference/generated/numpy.logical_and.html

```suggestion
    assert np.all((0 <= np_values) & (np_values <= 1))
```

**Review by himkt — COMMENTED** [review]
The appearances of `_validate` should be replaced with `_validate_edf_values`. ☝️

File: tests/visualization_tests/test_edf.py [hunk 8]
```diff
@@ -11,6 +14,15 @@ def test_target_is_none_and_study_is_multi_obj() -> None:
         plot_edf(study)
 
 
+def _validate_edf_values(edf_values: Sequence[float]) -> None:
+    np_values = np.array(y_values)
```
Comment (nzw0301) [inline tests/visualization_tests/test_edf.py:18]:
```suggestion
    np_values = np.array(edf_values)
```

File: tests/visualization_tests/test_edf.py [hunk 9]
```diff
@@ -28,27 +40,36 @@ def test_plot_optimization_history(direction: str) -> None:
...
 
     # Test with a customized target value.
     study0 = create_study(direction=direction)
     study0.optimize(lambda t: t.suggest_float("x", 0, 5), n_trials=10)
     with pytest.warns(UserWarning):
         figure = plot_edf(study0, target=lambda t: t.params["x"])
-    assert len(figure.data) == 1
+    data = figure.data
+    _validate_edf_values(data[0]["y"])
+    assert len(data) == 1
```
Comment (HideakiImamura) [inline tests/visualization_tests/test_edf.py:66]: For consistency as above.
```suggestion
    _validate_edf_values(figure.data[0]["y"])
    assert len(figure.data) == 1
```

File: tests/visualization_tests/test_edf.py [hunk 10]
```diff
@@ -28,27 +40,36 @@ def test_plot_optimization_history(direction: str) -> None:
...
 
     # Test with a customized target name.
     study0 = create_study(direction=direction)
     study0.optimize(lambda t: t.suggest_float("x", 0, 5), n_trials=10)
     figure = plot_edf(study0, target_name="Target Name")
-    assert len(figure.data) == 1
+    data = figure.data
+    _validate_edf_values(data[0]["y"])
+    assert len(data) == 1
```
Comment (HideakiImamura) [inline tests/visualization_tests/test_edf.py:74]: For consistency as above.
```suggestion   
     _validate_edf_values(figure.data[0]["y"])
    assert len(figure.data) == 1
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have two small comments. Could you take a look?

**Review by nzw0301 — APPROVED** [review]
LGTM. Thank you for improving the test!

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your updates. LGTM!

**Comment by toshihikoyanase** [comment]
@himkt @HideakiImamura This PR got two approvals and can I merge it?

**Comment by HideakiImamura** [comment]
@toshihikoyanase Thanks for the notification. Yes, I agree with the merge. Let me merge this PR.

**Pull request merged**